### PR TITLE
Test/resource mgt

### PR DESCRIPTION
**Pull request by bacchuongdaungo**
testcases for basic functionality and edge cases

### CODE REVIEW

File: src/EStimLibrary/Core/ReusableIdPool.cs [hunk 1]
```diff
@@ -44,6 +46,16 @@ public SortedSet<int> Ids
     {
         get
         {
+            if (this.NumIds == 0)
+            {
+                return new SortedSet<int>();
+            }
+
+            // Ensure that baseId + numIds - 1 does not exceed int.MaxValue
+            if ((long)this.BaseId + (long)this.NumIds - 1 > int.MaxValue)
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ReusableIdPool.cs]: Why are these (long) conversions?

File: src/EStimLibrary/Core/ReusableIdPool.cs [hunk 2]
```diff
@@ -102,13 +122,19 @@ public ReusableIdPool()
 
     public int IncrementNumIds(int increment)
     {
-        this.NumIds += increment;
-        return this.NumIds;
+        if (this.NumIds + increment <= int.MaxValue)
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ReusableIdPool.cs]: BUG: should not allow "numIds" to go below the current highest Used ID. This can be another part of the NumIds set() method

File: src/EStimLibrary/Core/ReusableIdPool.cs [hunk 3]
```diff
@@ -102,13 +122,19 @@ public ReusableIdPool()
...
+            this.NumIds += increment;
+            return this.NumIds;
+        }
+            return -1;
     }
 
     public int ResetNumIds(int newMax)
     {
-        this.NumIds = newMax;
+        if (newMax < 0) this.NumIds = 0;
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ReusableIdPool.cs]: Same comments as on the IncrementNumIds method

File: src/EStimLibrary/Core/ReusableIdPool.cs [hunk 4]
```diff
@@ -74,16 +86,24 @@ public SortedSet<int> FreeIds
     /// negative number is given. Can be changed after construction.</param>
     public ReusableIdPool(int baseId, int numIds)
     {
-        // Store the base ID. Property setter handles defaulting to 0.
-        this.BaseId = baseId;
+        // Ensure BaseId is not negative
+        this.BaseId = Math.Max(baseId, MIN_BASE_ID);
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ReusableIdPool.cs]: this check is already done via the protected set() method of the BaseId property (see above, around line 41), so you can remove it here

File: src/EStimLibrary/Core/ReusableIdPool.cs [hunk 5]
```diff
@@ -74,16 +86,24 @@ public SortedSet<int> FreeIds
...
 
-        // Store the number of IDs in this pool. Property setter handles
-        // defaulting to 0.
-        this.NumIds = numIds;
+        // Calculate the maximum allowable NumIds
+        long maxAllowableNumIds = (long)int.MaxValue - this.BaseId + 1;
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ReusableIdPool.cs]: Same question: why using long data type?

File: src/EStimLibrary/Core/ReusableIdPool.cs [hunk 6]
```diff
@@ -74,16 +86,24 @@ public SortedSet<int> FreeIds
...
 
-        // Initialize set of used IDs to empty. This is the only set manually
-        // that is method-editable not auto-generated.
-        this.UsedIds = new();
+        // Adjust NumIds if necessary
+        if ((long)numIds > maxAllowableNumIds)
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ReusableIdPool.cs]: Nice catch! Move this check to the NumIds property set() method so any time the value is attempted to be adjusted, this check is performed

File: src/EStimLibrary/Core/ReusableIdPool.cs (hunk 2 above)
Comment (lhmcgann) [reply src/EStimLibrary/Core/ReusableIdPool.cs]: Once have updated the NumIds set() method with both adjustments (min and max bound checking), can remove this if-else and simply edit NumIds and then return its value from the method

File: src/EStimLibrary/Core/ReusableIdPool.cs [hunk 7]
```diff
@@ -102,13 +122,19 @@ public ReusableIdPool()
...
+        }
+            return -1;
     }
 
     public int ResetNumIds(int newMax)
     {
-        this.NumIds = newMax;
+        if (newMax < 0) this.NumIds = 0;
+        else this.NumIds = newMax;
+        this.UsedIds.RemoveWhere(id => !this.IsValidId(id));
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ReusableIdPool.cs]: bound NumIds by the highest current used ID rather than deleting used IDs, so once NumIds set() method changes implemented, can delete this line

File: src/EStimLibrary/Core/ReusableIdPool.cs [hunk 8]
```diff
@@ -210,11 +237,15 @@ public bool TryGetNextFreeId(out int id)
 
     public SortedSet<int> GetSubset(int startId, int numIds)
     {
+        if (numIds <= 0)
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ReusableIdPool.cs:286]: Good case to think about: is it already covered by the way the rest of the method is carried out? Or was the GetViewBetween(x, x) giving you errors?

File: src/EStimLibrary/Core/ReusableIdPool.cs (hunk 2 above)
Comment (lhmcgann) [reply src/EStimLibrary/Core/ReusableIdPool.cs]: Once both of these changes are made, add test cases for them

File: src/EStimLibrary/Core/ReusableIdPool.cs [hunk 9]
```diff
@@ -162,6 +188,7 @@ public bool UseId(int globalId)
     /// </returns>
     public bool FreeId(int globalId)
     {
+        if (!this.IsValidId(globalId)) return false; // checks if the globalId is in range of the pool IDs
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ReusableIdPool.cs]: Nice catch!

File: src/EStimLibrary/Core/ResourceManager.cs [hunk 10]
```diff
@@ -8,13 +8,15 @@ public class ResourceManager<ResourceType>
 
     public readonly int MaxNumResources;
 
-    public int NumTotalResources => this.IdPool.NumUsedIds;
+    public int NumTotalResources => this.Resources.Count;
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ResourceManager.cs:11]: Was the previous implementation of this property failing? If so, something is broken, because in this case, IdPool.NumUsedIds and Resources.Count should always be the same

File: src/EStimLibrary/Core/ResourceManager.cs [hunk 11]
```diff
@@ -8,13 +8,15 @@ public class ResourceManager<ResourceType>
...
 
+    // Constructor
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ResourceManager.cs:13]: Make this a docstring comment, i.e., ///, etc
Also, include docstring comments for all classes, properties, fields, and methods

File: src/EStimLibrary/Core/ResourceManager.cs [hunk 12]
```diff
@@ -8,13 +8,15 @@ public class ResourceManager<ResourceType>
...
     public ResourceManager(int baseId = 0, int initialNumResourceIds = 0,
         int maxNumResources = Constants.POS_INFINITY)
     {
         this.IdPool = new(baseId, initialNumResourceIds);
         this.Resources = new();
+        if (maxNumResources < 0) maxNumResources = Constants.POS_INFINITY;
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ResourceManager.cs:19]: Ah I forgot I made a lib const for max int value later. Use Constants.POS_INFINITY in ReusableIdPool to wherever int.MaxValue is used right not

File: src/EStimLibrary/Core/ResourceManager.cs [hunk 13]
```diff
@@ -29,7 +31,7 @@ public bool IsValidResourceId(int globalId)
     {
         // TODO: does Used check validity again?
         return this.IdPool.IsValidId(globalId) &&
-            this.IdPool.IsUsed(globalId);
+            this.IdPool.IsUsed(globalId) && this.Resources.ContainsKey(globalId);
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ResourceManager.cs]: Is this line needed? Was it failing before? the IdPool.IsValidID + IsUsed checks should be equivalent, so if it was failing before, something bigger is broken

File: src/EStimLibrary/Core/ResourceManager.cs [hunk 14]
```diff
@@ -40,9 +42,9 @@ public bool TryGetNextAvailableId(out int globalId)
         while (!this.IdPool.TryGetNextFreeId(out globalId))
         {
             // If max capacity already used, return failure.
-            if (!Utils.IsWithinUpperBound(this.IdPool.NumUsedIds,
-                this.MaxNumResources))
+            if (this.IdPool.NumUsedIds >= this.MaxNumResources)
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ResourceManager.cs]: was the previous implementation failing? If so, should fix that rather than change a line here

File: src/EStimLibrary/Core/ResourceManager.cs [hunk 15]
```diff
@@ -63,7 +65,7 @@ public bool TryAddResource(int globalId, ResourceType resource)
         this.Resources.Add(globalId, resource);
         // small TODO: the Use method calls IsValidId again --> any way to
         // refactor to make more efficient?
-        this.IdPool.UseId(globalId);
+        this.IdPool.UsedIds.Add(globalId);
```
Comment (lhmcgann) [inline src/EStimLibrary/Core/ResourceManager.cs]: Was UseId() not adding it correctly? Similar thing: if that wasn't working, should fix that rather than work around it here. Or was there a different motivation for the change here?

**Review by lhmcgann — CHANGES_REQUESTED** [review]
I didn't look as closely at the specific test cases or asserts yet, but will do so after edits are made for the few comments left in this review. Also, if you can doc-comment all your test classes and methods, as well as all the library classes, methods, properties, etc they're testing, that would be great! Some examples in the commit I added to this branch

File: src/EStimLibrary/Core/ReusableIdPool.cs (hunk 1 above)
Comment (bacchuongdaungo) [reply src/EStimLibrary/Core/ReusableIdPool.cs]: because I was comparing the sum with int.MaxValue. If the sum are potentially larger than the int type can handle, then converting BaseId and NumIds to long makes sure we can compare.

File: src/EStimLibrary/Core/ReusableIdPool.cs (hunk 2 above)
Comment (bacchuongdaungo) [reply src/EStimLibrary/Core/ReusableIdPool.cs]: When you say NumIds set() method, do you mean this:
/// <summary>
/// The number of IDs in this pool. Cannot be below 0.
/// </summary>
public int NumIds
{
    get => this._numIds;
    protected set => this._numIds = Math.Max(MIN_NUM_IDS, value);
}

If so, does that mean that I can change the "set" functionality inside?

Comment (bacchuongdaungo) [reply src/EStimLibrary/Core/ReusableIdPool.cs]: I have pushed a commit fixing this. Please check if my modification is good

File: src/EStimLibrary/Core/ResourceManager.cs (hunk 13 above)
Comment (bacchuongdaungo) [reply src/EStimLibrary/Core/ResourceManager.cs]: I had a misunderstanding of the code. I mistook the id being used will make it invalid while in contrast it actually verifies the validity. I used Assert.False instead of True. I have fixed the test case

File: src/EStimLibrary/Core/ResourceManager.cs (hunk 15 above)
Comment (bacchuongdaungo) [reply src/EStimLibrary/Core/ResourceManager.cs]: I have reverted the changes and it passes the test cases

File: src/EStimLibrary/Core/ResourceManager.cs (hunk 14 above)
Comment (bacchuongdaungo) [reply src/EStimLibrary/Core/ResourceManager.cs]: I understand it now! The implementation of IsWithinUpperBound had "(inclusive)", but according to chatGPT, doing so results in Operator Precedence Issue, which would prioritize like so: "The ternary operator ? : has higher precedence than the logical OR ||.". So, I added parentheses in the return statement of IsWithinUpperBound, and added "inclusive: false" in TryGetNextAvailableId. It passed the test cases

File: src/EStimLibrary/Core/ResourceManager.cs (hunk 13 above)
Comment (bacchuongdaungo) [reply src/EStimLibrary/Core/ResourceManager.cs]: For the current implementation of IsValidId and IsUsed, it is crucial to have the line since we are checking for the resource. I have tried to implement the check in UseId but i can't do that without initiating an instance of ResourceManager, so I think it's simpler to have the check inside the RMgr class.
